tests/test_external_datasets.py

```diff
@@ -0,0 +1,241 @@
+        # wraps around:
+        assert generated_rows.row_values(7, "FirstName") == "Test"
+        assert generated_rows.row_values(7, "LastName") == "User"
+        assert generated_rows.row_values(8, "FirstName") == "Zest"
+
+    def test_csv_dataset_permutation(self, generated_rows):
+        abs_path = str(Path(__file__).parent)
+        yaml = (
+            """
+        - plugin: snowfakery.standard_plugins.datasets.Dataset
+        - object: XXX
+          count: 14
+          fields:
+            __address_from_csv:
+              Dataset.shuffle:
+                dataset: %s/../examples/datasets/addresses.csv
+            City: ${{__address_from_csv.City}}
+        """
+            % abs_path
+        )
+        generate(StringIO(yaml), {})
+        first_3 = [generated_rows.row_values(i, "City") for i in range(0, 3)]
+        assert len(first_3) == len(set(first_3))  # 3 unique items, in some order
+        next_3 = [generated_rows.row_values(i, "City") for i in range(3, 14)]
+        assert set(first_3) == set(next_3)
+
+    def test_SQL_dataset_permutation(self, generated_rows):
+        abs_path = str(Path(__file__).parent)
+        yaml = (
+            """
+        - plugin: snowfakery.standard_plugins.datasets.Dataset
+        - object: XXX
+          count: 14
+          fields:
+            __name_from_db:
+              Dataset.shuffle:
+                dataset: sqlite:///%s/databases/test_db.db
+                table: contacts
+            FirstName: ${{__name_from_db.first_name}}
+            LastName: ${{__name_from_db.last_name}}
+        """
+            % abs_path
+        )
+        generate(StringIO(yaml), {})
+        first_7 = [generated_rows.row_values(i, "FirstName") for i in range(0, 7)]
+        assert len(first_7) == len(set(first_7))  # 7 unique items, in some order
+        next_7 = [generated_rows.row_values(i, "FirstName") for i in range(7, 14)]
+        assert set(first_7) == set(next_7)
+
+    def test_SQL_dataset_permutation_really_shuffles(self, generated_rows):
+        abs_path = str(Path(__file__).parent)
+        yaml = (
+            """
+        - plugin: snowfakery.standard_plugins.datasets.Dataset
+        - object: XXX
+          count: 14
+          fields:
+            __name_from_db:
+              Dataset.shuffle:
+                dataset: sqlite:///%s/databases/test_db.db
+                table: contacts
+            FirstName: ${{__name_from_db.first_name}}
+            LastName: ${{__name_from_db.last_name}}
+        """
+            % abs_path
+        )
+
+        orig_query = SQLDatasetRandomPermutationIterator.query
+
+        called = None
+
+        def new_query(self):
+            nonlocal called
+            called = True
+            return orig_query(self)
+
+        with mock.patch(
+            "snowfakery.standard_plugins.datasets.SQLDatasetRandomPermutationIterator.query",
+            new_query,
+        ):
+            generate(StringIO(yaml), {})
+        assert called
```

Comment:
I think you could use mock's `wraps` option here: https://stackoverflow.com/a/25608441/212493

Reply:
The SO recipe requires me to create a SQLDatasetRandomPermutationIterator which takes two arguments which I'd need to create or mock.

And the recipe has an issue described in the comments which is that the method which is called is bound to a different to the dummy Test object rather than the one created in the code. So i guess it would be querying the wrong database?

And a naive wraps() is not sufficient because Mock objects can't be used as unbound methods. Python won't bind them because it only binds "real" functions. Other callable objects don't get bound.